pkg/apis/core/v1beta1/defaults_shoot.go

```diff
@@ -0,0 +1,385 @@
+// Copyright 2023 SAP SE or an SAP affiliate company. All rights reserved. This file is licensed under the Apache Software License, v. 2 except as noted otherwise in the LICENSE file
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package v1beta1
+
+import (
+	math "math"
+	"time"
+
+	v1beta1constants "github.com/gardener/gardener/pkg/apis/core/v1beta1/constants"
+	"github.com/gardener/gardener/pkg/utils/timewindow"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/utils/pointer"
+)
+
+// SetDefaults_Shoot sets default values for Shoot objects.
+func SetDefaults_Shoot(obj *Shoot) {
+	if obj.Spec.Kubernetes.KubeAPIServer == nil {
+		obj.Spec.Kubernetes.KubeAPIServer = &KubeAPIServerConfig{}
+	}
+	if obj.Spec.Kubernetes.KubeControllerManager == nil {
+		obj.Spec.Kubernetes.KubeControllerManager = &KubeControllerManagerConfig{}
+	}
```

Comment:
It is not clear for us when we extract the logic into a `SetDefaults_` func? For example why the above handing is not extracted into a `SetDefaults_Kubernetes` func? Same for `SetDefaults_Maintenance`?
Can we also document the motivation for https://github.com/gardener/gardener/issues/7312 somewhere in our docs so that the motivation does not only stay in an issue only.

Reply:
The issue is `Kubernetes` defaulting is needed only in case of shoot with workers. How the `zz_generated.defaults.go` file will check for the workerless condition for shoot?

Reply:
Ok, I see.

Reply:
> Can we also document the motivation for #7312 somewhere in our docs so that the motivation does not only stay in an issue only.

This part is not addressed. That's why I added a new subtask to https://github.com/gardener/gardener/issues/7312.